File: day01/src/main.py

def get_input_data():
    with open('../input.txt') as f:
        lines = [s.strip('\n') for s in f.readlines()]
    lefties = []
    righties = []

    for line in lines:
        left, right = line.split()
        lefties.append(int(left))
        righties.append(int(right))
    return lefties, righties


def part1():
    lefties, righties = get_input_data()
    lefties.sort()
    righties.sort()
    total = 0

    for lhs, rhs in zip(lefties, righties):
        total += abs(lhs - rhs)

    print(f'Part 1 {total}')


def part2():
    lefties, righties = get_input_data()
    total = 0
    for num in lefties:
        total += num * righties.count(num)
    print(f'Part 2 {total}')


def main():
    part1()
    part2()


if __name__ == '__main__':
    main()

File: day01/src/main.rs
use aoc::get_input_as_lines;
use itertools::Itertools;
use std::convert::TryFrom;


fn parse_input() -> (Vec<i32>, Vec<i32>){
    let lines: Vec<&str> = get_input_as_lines(include_str!("../example.txt"));
    let mut left_nums: Vec<i32> = Vec::new();
    let mut right_nums: Vec<i32> = Vec::new();
    for line in lines {
        let nums: Vec<i32> = line.split("   ").map(|x|->i32{x.parse().unwrap()}).collect();
        left_nums.push(nums.clone()[0]);
        right_nums.push(nums.clone()[1]);
    }
    left_nums.sort();
    right_nums.sort();
    (left_nums, right_nums)
}

fn part1() {
    let mut total_distance: u32 = 0;
    let (left_nums, right_nums) = parse_input();
    for (left, right) in left_nums.iter().zip(right_nums.iter()){
        total_distance += left.abs_diff(*right);
    }
    println!("Part 1: {}", total_distance);
}

fn part2() {
    let (left_nums, right_nums) = parse_input();
    let mut right_counts: std::collections::HashMap<i32, usize> = right_nums.into_iter().counts();
    let mut total_count: usize = 0;
    for num in left_nums {
        total_count += usize::try_from(num).ok().unwrap() * *right_counts.entry(num).or_insert(0);
    }
    println!("Part 2: {}", total_count);
}


fn main() {
    part1();
    part2();
}